### PR TITLE
FEAT-#6735: Make Modin on MPI through unidist component more obvious

### LINKED ISSUE #6735: Make Modin on MPI through unidist component more obvious
We should make Modin on MPI through unidist component more obvious to emphasize MPI backend for Modin.

### PR DESCRIPTION
**Pull request by YarShev**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6735  
- [x] tests passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

**Comment by YarShev** [comment]
Blocked for merging but open for review. Once unidist 0.5.0 is available both on PyPI and conda-forge, I will rebase on master to run CI again to check if tests pass.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -97,7 +97,6 @@ jobs:
         run: |
           MODIN_ENGINE=dask python -c "import modin.pandas as pd; print(pd.DataFrame([1,2,3]))"
           MODIN_ENGINE=ray python -c "import modin.pandas as pd; print(pd.DataFrame([1,2,3]))"
-          MODIN_ENGINE=unidist UNIDIST_BACKEND=mpi mpiexec -n 1 python -c "import modin.pandas as pd; print(pd.DataFrame([1,2,3]))"
```
Comment (anmyachev) [inline .github/workflows/ci.yml:100]: We still need to test the clean install of unidist using pip.

File: examples/tutorial/jupyter/execution/pandas_on_unidist/Dockerfile [hunk 2]
```diff
@@ -1,5 +1,4 @@
 FROM continuumio/miniconda3
 
+RUN conda env create -f jupyter_unidist_env.yml
```
Comment (anmyachev) [inline examples/tutorial/jupyter/execution/pandas_on_unidist/Dockerfile:3]: Why did you change the installation method?

File: docs/flow/modin/core/execution/unidist/implementations/pandas_on_unidist/index.rst [hunk 3]
```diff
@@ -6,7 +6,8 @@ PandasOnUnidist Execution
 Queries that perform data transformation, data ingress or data egress using the `pandas on Unidist` execution
 pass through the Modin components detailed below.
 
-To enable `pandas on Unidist` execution, please refer to the usage section in :doc:`pandas on Unidist </development/using_pandas_on_unidist>`.
+To enable `pandas on MPI through unidist` execution,
```
Comment (anmyachev) [inline docs/flow/modin/core/execution/unidist/implementations/pandas_on_unidist/index.rst:9]: Such changes in documentation can be made independently from the new release of the unidist, right?

**Comment by anmyachev** [comment]
@YarShev with the new Unidist release, all of Modin’s previous releases on Unidist break (at least installation with pip). I only know a way to fix it using an explicit unidist pin: `<0.0.5`. Do you know a better solution?

**Comment by YarShev** [comment]
> @YarShev with the new Unidist release, all of Modin’s previous releases on Unidist break (at least installation with pip). I only know a way to fix it using an explicit unidist pin: `<0.5.0`. Do you know a better solution?

@anmyachev, I want to say that this is not a bug. This is a feature, which allows to use more recent MPICH and Python versions, as well as use other MPI implementations to run on. As of unidist 0.5.0 it is required to have a working MPI implementation installed beforehand. As far as I can tell, there is no anything we can do with the previous releases of `modin-unidist` working out of the box other than 1) specify the exact version of unidist when installing (`pip install modin-unidist unidist-mpi<=0.4.1`) or 2) wait for mpi4py releases on PyPI that have a pre-built MPI implementation and thus work out of the box.

Note: we are talking only about pip.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (YarShev) [reply .github/workflows/ci.yml:100]: Added for ubuntu.

File: docs/flow/modin/core/execution/unidist/implementations/pandas_on_unidist/index.rst (hunk 3 above)
Comment (YarShev) [reply docs/flow/modin/core/execution/unidist/implementations/pandas_on_unidist/index.rst:9]: Right. unidist 0.5.0 is out so it is okay anyway.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (anmyachev) [reply .github/workflows/ci.yml:100]: Why also not for Windows?

File: setup.py [hunk 4]
```diff
@@ -9,9 +9,13 @@
...
 ray_deps = ["ray[default]>=1.13.0,!=2.5.0", "pyarrow>=7.0.0", "pydantic<2"]
-unidist_deps = ["unidist[mpi]>=0.2.1"]
+mpi_deps = ["unidist[mpi]>=0.2.1"]
 spreadsheet_deps = ["modin-spreadsheet>=0.1.0"]
-all_deps = dask_deps + ray_deps + unidist_deps + spreadsheet_deps
+# Currently, Modin does not include `mpi` option in `all`.
+# Otherwise, installation of modin[all] would fail because
+# users need to have a working MPI implementation and
+# certain software installed beforehand.
+all_deps = dask_deps + ray_deps + spreadsheet_deps
```
Comment (anmyachev) [inline setup.py:18]: I understand your motivation, but it confuses me that the target, which is called like `all`, is not that in essence.

**Review by anmyachev — APPROVED** [review]
LGTM!

File: setup.py (hunk 4 above)
Comment (YarShev) [reply setup.py:18]: We have a similar issue with hdk. I don't think we want users to see the installation error with new releases of modin[all] for now. Maybe we need some time for users to understand the behavior/logic of modin[mpi].

File: .github/workflows/ci.yml (hunk 1 above)
Comment (YarShev) [reply .github/workflows/ci.yml:100]: There is a little bit more stuff to install for MPI on Windows. I don't even know if it is possible to install all that stuff without interacting with GUI. Originally, Modin tests for unidist on MPI were added only on Linux so I think it is okay to leave the things here on Linux too.

Comment (anmyachev) [reply .github/workflows/ci.yml:100]: Ok, just curious: maybe it’s enough to just get the prerequisites for MPI through conda?

Comment (YarShev) [reply .github/workflows/ci.yml:100]: It should be enough but in that case we would need to switch to `conda-incubator/setup-miniconda@v2` usage here, which I didn't want to.

**Comment by anmyachev** [comment]
@YarShev CI is failed

**Comment by YarShev** [comment]
> @YarShev CI is failed

I see, looking into this.

File: setup.py (hunk 4 above)
Comment (dchigarev) [reply setup.py:18]: can we add a logic, checking if any supported `mpi` version is installed and then either install `unidist` or raise a warning about skipping it?

Comment (YarShev) [reply setup.py:18]: Just checking if mpi is available on the system is not enough. There are much more other stuff that is important and required for mpi4py to be installed. mpi4py itself has the appropriate logic for observing an MPI library/compiler/etc. and I believe this should belong to mpi4py. To me it is fine to go with modin[mpi] for now.

**Review by anmyachev — APPROVED** [review]
LGTM!

**Comment by YarShev** [comment]
Rebased just to rerun CI.
